sklearn/isotonic.py

```diff
@@ -317,11 +343,24 @@ def fit(self, X, y, sample_weight=None):
         self._build_y(X, y, sample_weight)
 
         # Handle the left and right bounds on X
-        self.X_min_ = np.min(self.X_)
-        self.X_max_ = np.max(self.X_)
+        self.X_min_ = np.min(self._X_)
+        self.X_max_ = np.max(self._X_)
+
+        # Remove unnecessary points for faster prediction
+        keep_data = np.ones((len(self._y_),), dtype=bool)
+        # Aside from the 1st and last point, remove points whose y values
+        # are equal to both the point before and the point after it.
+        keep_data[1:-1] = np.logical_or(
+            np.not_equal(self._y_[1:-1], self._y_[:-2]),
+            np.not_equal(self._y_[1:-1], self._y_[2:])
+            )
+        # We're keeping self.X_ and self.y_ around for backwards compatibility,
+        # but they should be considered deprecated.
+        self._necessary_X_ = self._X_[keep_data]
+        self._necessary_y_ = self._y_[keep_data]
```

Comment:
I think  this block of code should be moved into the `_build_y` function and slightly refactored so that it does not rely upon the deprecated `self._y_` attribute.